File: src/gx/utils/git.py
"""Shared git helpers used across gx commands."""

from __future__ import annotations

import subprocess
from datetime import datetime, timezone


class GitError(Exception):
    """Raised when a git command fails."""


def run_git(args: list[str], cwd: str | None = None, check: bool = True) -> str:
    """Run a git command and return stdout. Raises GitError on failure if check=True."""
    try:
        result = subprocess.run(
            ["git"] + args,
            capture_output=True,
            text=True,
            cwd=cwd,
            timeout=30,
        )
        if check and result.returncode != 0:
            stderr = result.stderr.strip()
            raise GitError(stderr or f"git {' '.join(args)} failed with code {result.returncode}")
        return result.stdout.strip()
    except FileNotFoundError:
        raise GitError("git is not installed or not in PATH")
    except subprocess.TimeoutExpired:
        raise GitError(f"git {' '.join(args)} timed out after 30 seconds")


def ensure_git_repo(cwd: str | None = None) -> None:
    """Raise GitError if not inside a git repository."""
    try:
        run_git(["rev-parse", "--is-inside-work-tree"], cwd=cwd)
    except GitError:
        raise GitError("Not a git repository. Run this from inside a git project.")


def get_current_branch(cwd: str | None = None) -> str:
    """Return the current branch name. Raises GitError if in detached HEAD."""
    branch = run_git(["symbolic-ref", "--short", "HEAD"], cwd=cwd, check=False)
    if not branch:
        raise GitError("HEAD is detached. Not on any branch.")
    return branch


def get_head_branch(cwd: str | None = None) -> str:
    """Detect the repo's HEAD branch (main, master, develop, etc.)."""
    # Strategy 1: check remote HEAD
    try:
        output = run_git(["remote", "show", "origin"], cwd=cwd, check=False)
        for line in output.splitlines():
            if "HEAD branch:" in line:
                return line.split("HEAD branch:")[-1].strip()
    except GitError:
        pass

    # Strategy 2: check common branch names
    for name in ("main", "master", "develop"):
        if branch_exists(name, cwd=cwd):
            return name

    # Strategy 3: git config
    try:
        default = run_git(["config", "--get", "init.defaultBranch"], cwd=cwd, check=False)
        if default and branch_exists(default, cwd=cwd):
            return default
    except GitError:
        pass

    return "main"


def get_repo_root(cwd: str | None = None) -> str:
    return run_git(["rev-parse", "--show-toplevel"], cwd=cwd)


def is_clean_working_tree(cwd: str | None = None) -> bool:
    output = run_git(["status", "--porcelain"], cwd=cwd)
    return len(output) == 0


def get_last_commit(cwd: str | None = None) -> dict:
    fmt = "%H%n%h%n%s%n%an%n%aI"
    try:
        output = run_git(["log", "-1", f"--format={fmt}"], cwd=cwd)
    except GitError:
        return {}
    lines = output.splitlines()
    if len(lines) < 5:
        return {}
    return {
        "hash": lines[0],
        "short_hash": lines[1],
        "message": lines[2],
        "author": lines[3],
        "date": lines[4],
    }


def get_stash_count(cwd: str | None = None) -> int:
    output = run_git(["stash", "list"], cwd=cwd, check=False)
    if not output:
        return 0
    return len(output.splitlines())


def branch_exists(name: str, remote: bool = False, cwd: str | None = None) -> bool:
    if remote:
        result = run_git(["branch", "-r", "--list", f"origin/{name}"], cwd=cwd, check=False)
    else:
        result = run_git(["branch", "--list", name], cwd=cwd, check=False)
    return len(result.strip()) > 0


def get_merge_base(branch_a: str, branch_b: str, cwd: str | None = None) -> str:
    return run_git(["merge-base", branch_a, branch_b], cwd=cwd)


def get_reflog_entries(n: int = 20, cwd: str | None = None) -> list[dict]:
    fmt = "%H%n%gd%n%gs%n%gD%n%ci"
    try:
        output = run_git(["reflog", f"--format={fmt}", "-n", str(n)], cwd=cwd)
    except GitError:
        return []
    entries = []
    lines = output.splitlines()
    i = 0
    while i + 4 < len(lines):
        entries.append({
            "hash": lines[i],
            "ref": lines[i + 1],
            "action": lines[i + 2],
            "description": lines[i + 3],
            "timestamp": lines[i + 4],
        })
        i += 5
    return entries


def is_commit_pushed(ref: str = "HEAD", cwd: str | None = None) -> bool:
    """Check if a commit exists on the remote tracking branch."""
    try:
        branch = get_current_branch(cwd=cwd)
        tracking = run_git(
            ["rev-parse", "--abbrev-ref", f"{branch}@{{upstream}}"],
            cwd=cwd,
            check=False,
        )
        if not tracking:
            return False
        result = run_git(
            ["branch", "-r", "--contains", ref],
            cwd=cwd,
            check=False,
        )
        return len(result.strip()) > 0
    except GitError:
        return False


def time_ago(date_str: str) -> str:
    """Convert an ISO date string to a human-readable relative time."""
    try:
        dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
        now = datetime.now(timezone.utc)
        diff = now - dt
        seconds = int(diff.total_seconds())
        if seconds < 60:
            return "just now"
        minutes = seconds // 60
        if minutes < 60:
            return f"{minutes} min ago"
        hours = minutes // 60
        if hours < 24:
            return f"{hours} hour{'s' if hours != 1 else ''} ago"
        days = hours // 24
        if days < 30:
            return f"{days} day{'s' if days != 1 else ''} ago"
        months = days // 30
        return f"{months} month{'s' if months != 1 else ''} ago"
    except (ValueError, TypeError):
        return date_str

File: internal/git/git.go
// Package git provides helpers for executing git commands.
package git

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"sync"
	"time"
)

// Git state file paths (relative to .git/).
const (
	MergeHeadPath      = "MERGE_HEAD"
	RebaseMergePath    = "rebase-merge"
	RebaseApplyPath    = "rebase-apply"
	CherryPickHeadPath = "CHERRY_PICK_HEAD"
)

// Error is returned when a git command fails.
type Error struct {
	Args   []string
	Stderr string
	Err    error
}

func (e *Error) Error() string {
	if e.Stderr != "" {
		return e.Stderr
	}
	return fmt.Sprintf("git %s failed: %v", strings.Join(e.Args, " "), e.Err)
}

// Run executes a git command and returns trimmed stdout.
func Run(args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return "", &Error{Args: args, Stderr: strings.TrimSpace(stderr.String()), Err: err}
	}
	return strings.TrimSpace(stdout.String()), nil
}

// RunCombined executes a git command and returns combined stdout+stderr.
// Used for rebase operations that need to capture conflict output.
func RunCombined(args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	out, err := cmd.CombinedOutput()
	output := strings.TrimSpace(string(out))
	if err != nil {
		return output, &Error{Args: args, Stderr: output, Err: err}
	}
	return output, nil
}

// RunUnchecked executes a git command and returns stdout even on failure.
func RunUnchecked(args ...string) string {
	out, _ := Run(args...)
	return out
}

// Lines executes a git command and returns stdout split into lines.
func Lines(args ...string) ([]string, error) {
	output, err := Run(args...)
	if err != nil {
		return nil, err
	}
	if output == "" {
		return nil, nil
	}
	return strings.Split(output, "\n"), nil
}

// EnsureRepo checks that we are inside a git repository.
func EnsureRepo() error {
	_, err := Run("rev-parse", "--is-inside-work-tree")
	if err != nil {
		return fmt.Errorf("not a git repository. Run this from inside a git project")
	}
	return nil
}

// CurrentBranch returns the current branch name.
func CurrentBranch() (string, error) {
	branch, err := Run("symbolic-ref", "--short", "HEAD")
	if err != nil || branch == "" {
		return "", fmt.Errorf("HEAD is detached. Not on any branch")
	}
	return branch, nil
}

var (
	headBranchOnce  sync.Once
	headBranchCache string
)

// HeadBranch detects the repo's main/master/develop branch.
// The result is cached for the lifetime of the process.
func HeadBranch() string {
	headBranchOnce.Do(func() {
		headBranchCache = detectHeadBranch()
	})
	return headBranchCache
}

func detectHeadBranch() string {
	// Try local symbolic ref for origin HEAD (no network call)
	out := RunUnchecked("symbolic-ref", "refs/remotes/origin/HEAD")
	if out != "" {
		// Output is e.g. "refs/remotes/origin/main"
		name := strings.TrimPrefix(out, "refs/remotes/origin/")
		if name != "" && BranchExists(name) {
			return name
		}
	}

	for _, name := range []string{"main", "master", "develop"} {
		if BranchExists(name) {
			return name
		}
	}

	// Last resort: use whatever branch HEAD is on
	if branch, err := CurrentBranch(); err == nil && branch != "" {
		return branch
	}
	return "main"
}

// FileExists checks if a file (not directory) exists at path.
func FileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

// DirExists checks if a directory exists at path.
func DirExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

// RepoRoot returns the root directory of the git repo.
func RepoRoot() (string, error) {
	return Run("rev-parse", "--show-toplevel")
}

// BranchExists checks if a local branch exists.
func BranchExists(name string) bool {
	_, err := Run("rev-parse", "--verify", name)
	return err == nil
}

// RemoteBranchExists checks if a remote branch exists.
func RemoteBranchExists(name string) bool {
	out := RunUnchecked("branch", "-r", "--list", "origin/"+name)
	return strings.TrimSpace(out) != ""
}

// IsClean returns true if the working tree has no uncommitted changes.
func IsClean() bool {
	out := RunUnchecked("status", "--porcelain")
	return out == ""
}

// LastCommit returns info about the HEAD commit.
func LastCommit() (hash, shortHash, message, author, date string) {
	out := RunUnchecked("log", "-1", "--format=%H%n%h%n%s%n%an%n%aI")
	lines := strings.Split(out, "\n")
	if len(lines) >= 5 {
		return lines[0], lines[1], lines[2], lines[3], lines[4]
	}
	return "", "", "", "", ""
}

// StashCount returns the number of stash entries.
func StashCount() int {
	out := RunUnchecked("stash", "list")
	if out == "" {
		return 0
	}
	return len(strings.Split(out, "\n"))
}

// MergeBase returns the merge base commit of two refs.
func MergeBase(a, b string) (string, error) {
	return Run("merge-base", a, b)
}

// AheadBehind returns the ahead/behind counts between two refs.
func AheadBehind(a, b string) (ahead, behind int) {
	out := RunUnchecked("rev-list", "--left-right", "--count", a+"..."+b)
	parts := strings.Fields(out)
	if len(parts) == 2 {
		fmt.Sscanf(parts[0], "%d", &ahead)
		fmt.Sscanf(parts[1], "%d", &behind)
	}
	return
}

// IsCommitPushed checks if HEAD exists on the remote tracking branch.
func IsCommitPushed() bool {
	branch, err := CurrentBranch()
	if err != nil {
		return false
	}
	tracking := RunUnchecked("rev-parse", "--abbrev-ref", branch+"@{upstream}")
	if tracking == "" {
		return false
	}
	out := RunUnchecked("branch", "-r", "--contains", "HEAD")
	return strings.TrimSpace(out) != ""
}

// TimeAgo converts an ISO date string to a human-readable relative time.
func TimeAgo(dateStr string) string {
	t, err := time.Parse(time.RFC3339, dateStr)
	if err != nil {
		// Try alternate format
		t, err = time.Parse("2006-01-02 15:04:05 -0700", dateStr)
		if err != nil {
			return dateStr
		}
	}
	diff := time.Since(t)
	switch {
	case diff < time.Minute:
		return "just now"
	case diff < time.Hour:
		m := int(diff.Minutes())
		return fmt.Sprintf("%d min ago", m)
	case diff < 24*time.Hour:
		h := int(diff.Hours())
		if h == 1 {
			return "1 hour ago"
		}
		return fmt.Sprintf("%d hours ago", h)
	case diff < 30*24*time.Hour:
		d := int(diff.Hours() / 24)
		if d == 1 {
			return "1 day ago"
		}
		return fmt.Sprintf("%d days ago", d)
	default:
		m := int(diff.Hours() / 24 / 30)
		if m <= 1 {
			return "1 month ago"
		}
		return fmt.Sprintf("%d months ago", m)
	}
}
